src/Voron/Impl/WeakSliceSmallSet.cs

```diff
@@ -0,0 +1,142 @@
+﻿using System;
+using System.Buffers;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using Sparrow;
+using Sparrow.Server;
+
+namespace Voron.Impl
+{
+    // The concept of the weak small set is about an optimized set that can lose elements when it gets filled.
+    // In a sense it behaves like a LRU cache, but will not track the read accesses. This version of the SmallSet
+    // is specially designed to deal with sequence values stored in byte pointers. The main difference with a
+    // dictionary is that the small set will calculate the hash ONLY if there are candidates that match already. 
+    public sealed class WeakSliceSmallSet<TValue> : IDisposable
+    {
+        private const int Invalid = -1;
+
+        private readonly int _length;
+        private readonly int[] _keySizes;
+        private readonly ulong[] _keyHashes;
+        private readonly Slice[] _keys;
+        private readonly TValue[] _values;
+        private int _currentIdx;
+
+        public WeakSliceSmallSet(int size = 0)
+        {
+            _length = size > Vector<long>.Count ? size - size % Vector<long>.Count : Vector<long>.Count;
+            _keySizes = ArrayPool<int>.Shared.Rent(_length);
+            _keyHashes = ArrayPool<ulong>.Shared.Rent(_length);
+            _keys = ArrayPool<Slice>.Shared.Rent(_length);
+            _values = ArrayPool<TValue>.Shared.Rent(_length);
+            _currentIdx = -1;
+        }
+
+        public unsafe void Add(Slice key, TValue value)
+        {
+            int idx = FindKey(key);
+            if (idx == Invalid)
```

Comment:
We can do a chain here, no? For the rare case that we have a collision, we'll put a value that means, use a backend dictionary for that?
Then we'll have the normal case, fast, and dictionary access in the next stage.

Reply:
Yes, but it is not the "Weak" implementation :) ... the idea of the weak set is to use it like a cache; we don't care if we cannot find the object there. In a sense it is like a `WeakReference` where you need to have a fallback in place if you end up not finding it. The sizes you pick are also big enough to guarantee that is a very uncommon situation while at the same time be fast.

Reply:
Everywhere you are using that, you actually _need_ those values, so you have another dictionary to back you up there.
It is an issue.

Reply:
No, we can get away with just going to retrieve it again. It will cost us a few extra calls like `GetPage` but that is about it.

Reply:
BTW, sizes have been selected to avoid that to happen altogether.

Reply:
Assuming we keep this, the problem is that you may have identity semantics (such as holding `Tree`, so that is a _big_ issue.

Reply:
99% certain that this will be used wrongly, however.
And the kind of bugs that come out of that are _nasty_.

Reply:
@redknightlois What is the status here?